observer/test/platform-stats.test.js

```diff
@@ -0,0 +1,71 @@
+import assert from 'node:assert'
+import pg from 'pg'
+import { beforeEach, describe, it } from 'mocha'
+
+import { DATABASE_URL } from '../lib/config.js'
+import { updateDailyTransferStats } from '../lib/platform-stats.js'
+import { migrateWithPgClient } from '@filecoin-station/spark-stats-db-migrations'
+
+describe('platform-stats-generator', () => {
+  /** @type {pg.Pool} */
+  let pgPool
+
+  before(async () => {
+    pgPool = new pg.Pool({ connectionString: DATABASE_URL })
+    await migrateWithPgClient(pgPool)
+  })
+
+  let today
+  beforeEach(async () => {
+    await pgPool.query('DELETE FROM daily_reward_transfers')
+
+    // Run all tests inside a transaction to ensure `now()` always returns the same value
+    // See https://dba.stackexchange.com/a/63549/125312
+    // This avoids subtle race conditions when the tests are executed around midnight.
+    await pgPool.query('BEGIN TRANSACTION')
```

Comment:
There is a subtle bug here: `pgPool.query` grabs a free (or opens a new connection), executes the query, and returns the connection back to the pool.

If we happen to run two queries at the same time, the first will run on the first connection inside this transaction, while the other will run outside of the transaction and therefore it will see different current date.

If you look at the tests in spark-evaluate, they are using a single connection (`pgClient`), and therefore they are not affected by this problem.
https://github.com/filecoin-station/spark-evaluate/blob/main/test/platform-stats.test.js

Here is an example test where we have both `pgPool` and `pgClient`:
https://github.com/filecoin-station/spark-api/blob/a18712008784725f6ebc98eda6a268558ba260b3/api/test/round-tracker.test.js#L24-L33

I propose few alternative ways how to fix this issue:

(1)
Do we really need `pg.Pool` in `updateDailyTransferStats`? Can we rework the code to work accept `pg.Client`? (Note that `pg.Pool` is a superset of `pg.Client`, so you can pass `pg.Pool` to arguments accepting a `pg.Client`).

IIRC, we were using `pg.Pool` in places where we were executing transactions and wanted to avoid a foot gun where a function would assume all queries are executed on the same connection, but the caller would provide a pool, and therefore, some of the queries would be executed on a different connection outside the transaction.

(2)
If we need `pg.Pool,` we need to find a different way to work with the current date. For example, we can call `await getCurrentDate()` in every test. Unfortunately, this is not bullet proof either - if the code under test executes the query just before midnight and the test calls `now()` after midnight, they will see a different day.

(3)
Since this problem manifests only around midnight UTC, I guess we can also ignore it. If we decide so, then let's remove the `BEGIN TRANSACTION` query to avoid a false sense of safety.

@PatrickNercessian @juliangruber WDYT?

Reply:
+1 to (1)